pkg/library/status/check.go

```diff
@@ -0,0 +1,167 @@
+func CheckPodsState(workspaceID string, namespace string, labelSelector k8sclient.MatchingLabels,
+	clusterAPI sync.ClusterAPI) (stateMsg string, checkFailure error) {
+	podList, err := GetPods(namespace, labelSelector, clusterAPI.Client)
+	if err != nil {
+		return "", err
+	}
+
+	for _, pod := range podList.Items {
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			if !CheckContainerStatusForFailure(&containerStatus) {
+				return fmt.Sprintf("Container %s has state %s", containerStatus.Name, containerStatus.State.Waiting.Reason), nil
+			}
+		}
+		for _, initContainerStatus := range pod.Status.InitContainerStatuses {
+			if !CheckContainerStatusForFailure(&initContainerStatus) {
+				return fmt.Sprintf("Init Container %s has state %s", initContainerStatus.Name, initContainerStatus.State.Waiting.Reason), nil
+			}
+		}
+		if msg, err := CheckPodEvents(&pod, workspaceID, clusterAPI); err != nil || msg != "" {
+			return msg, err
+		}
+	}
+	return "", nil
+}
+
+func CheckPodEvents(pod *corev1.Pod, workspaceID string, clusterAPI sync.ClusterAPI) (msg string, err error) {
+	evs := &corev1.EventList{}
+	selector, err := fields.ParseSelector(fmt.Sprintf("involvedObject.name=%s", pod.Name))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse field selector: %s", err)
+	}
+	if err := clusterAPI.Client.List(clusterAPI.Ctx, evs, k8sclient.InNamespace(pod.Namespace), k8sclient.MatchingFieldsSelector{Selector: selector}); err != nil {
+		return "", fmt.Errorf("failed to list events in namespace %s: %w", pod.Namespace, err)
+	}
+	for _, ev := range evs.Items {
+		if ev.InvolvedObject.Kind != "Pod" {
+			continue
+		}
+
+		// On OpenShift, it's possible see "FailedMount" events when using a routingClass that depends on the service-ca
+		// operator. To avoid this, we always ignore FailedMount events if the message refers to the DWO-provisioned volume
+		if infrastructure.IsOpenShift() &&
+			ev.Reason == "FailedMount" &&
+			strings.Contains(ev.Message, common.ServingCertVolumeName(common.ServiceName(workspaceID))) {
+			continue
+		}
+
+		if maxCount, isUnrecoverableEvent := unrecoverablePodEventReasons[ev.Reason]; isUnrecoverableEvent {
+			if !checkIfUnrecoverableEventIgnored(ev.Reason) && ev.Count >= maxCount {
+				var msg string
+				if ev.Count > 1 {
+					msg = fmt.Sprintf("Detected unrecoverable event %s %d times: %s.", ev.Reason, ev.Count, ev.Message)
+				} else {
+					msg = fmt.Sprintf("Detected unrecoverable event %s: %s.", ev.Reason, ev.Message)
+				}
+				return msg, nil
+			}
+		}
+	}
+	return "", nil
+}
+
+func CheckContainerStatusForFailure(containerStatus *corev1.ContainerStatus) (ok bool) {
+	if containerStatus.State.Waiting != nil {
+		for _, failureReason := range containerFailureStateReasons {
+			if containerStatus.State.Waiting.Reason == failureReason {
+				return checkIfUnrecoverableEventIgnored(containerStatus.State.Waiting.Reason)
+			}
+		}
+	}
+	return true
+}
+
+// TODO: Remove this function?
```

Comment:
You can drop this comment now that GetPods is more generic and used multiple places, though note that it's not really making life that much simpler:
```go
func GetPods(namespace string, labelSelector k8sclient.MatchingLabels, client runtimeClient.Client) (*corev1.PodList, error) {
	pods := &corev1.PodList{}
	if err := client.List(context.TODO(), pods, k8sclient.InNamespace(namespace), labelSelector); err != nil {
		return nil, err
	}
	return pods, nil
}
// ...
pods, err := check.GetPods(job.Namespace, k8sclient.MatchingLabels{"job-name": common.PVCCleanupJobName(workspaceID)}, clusterAPI.Client)
if err != nil {
  return "", err
}
```
versus
```go
pods := &corev1.PodList{}
if err := client.List(context.TODO(), pods, k8sclient.InNamespace(namespace), labelSelector); err != nil {
	return nil, err
}
```

IMO it's cleaner to just drop the function and do the listing where we need it, since it's not doing anything fancy -- most of the lines of code are just checking the error multiple times.

Reply:
Agreed, it doesn't make things more simple and the same amount of error checking is still required, so I'm dropping it.